File: test_crawler.js
#!/usr/bin/env node

/**
 * Test script for the Python Crawl4AI scraper
 * Run with: node test_crawler.js
 */

const { exec } = require('child_process');
const path = require('path');
const fs = require('fs');

// Path to the Python crawler script
const crawlerPath = path.join(__dirname, 'src', 'services', 'crawler.py');

// Test sources - modify as needed
const testSources = [
  { identifier: 'https://example.com', type: 'website' },
  // Add more test sources as needed
];

// Convert sources to JSON string for passing to Python
const sourcesJson = JSON.stringify(testSources);

console.log('Testing Python Crawl4AI crawler...');
console.log(`Sources: ${sourcesJson}`);

// Run the Python crawler with our test sources
exec(`python ${crawlerPath} '${sourcesJson}'`, (error, stdout, stderr) => {
  if (error) {
    console.error(`Error executing crawler: ${error.message}`);
    return;
  }
  
  if (stderr) {
    console.error(`Crawler stderr: ${stderr}`);
  }
  
  // Process the stdout as JSON
  try {
    const result = JSON.parse(stdout);
    
    console.log('\n--- Test Results ---');
    console.log(`Total stories found: ${result.stories?.length || 0}`);
    
    if (result.errors && result.errors.length > 0) {
      console.log('\nErrors encountered:');
      result.errors.forEach(error => {
        console.log(`- ${error.source}: ${error.error}`);
      });
    }
    
    if (result.stories && result.stories.length > 0) {
      console.log('\nSample stories:');
      result.stories.slice(0, 3).forEach((story, index) => {
        console.log(`\n${index + 1}. ${story.headline}`);
        console.log(`   Link: ${story.link}`);
        console.log(`   Date: ${story.date_posted}`);
      });
      
      // Save full results to a file for inspection
      fs.writeFileSync(
        'crawler_test_results.json', 
        JSON.stringify(result, null, 2),
        'utf8'
      );
      console.log('\nFull results saved to crawler_test_results.json');
    } else {
      console.log('\nNo stories found in test sources.');
    }
    
  } catch (parseError) {
    console.error('Error parsing crawler output:');
    console.error(parseError);
    console.log('\nRaw output:');
    console.log(stdout);
  }
});

File: test_crawler.py
#!/usr/bin/env python
"""
Simple test script for the crawler
"""

import sys
import json
import subprocess
import os

# Test sources
test_sources = [
    {"identifier": "https://example.com", "type": "website"}
]

# Convert to JSON string
sources_json = json.dumps(test_sources)

# Path to crawler script
crawler_path = os.path.join("src", "services", "crawler.py")

print(f"Testing crawler with: {sources_json}")

# Run the crawler with this input
try:
    # Use single quotes for the JSON string on Windows
    cmd = f"python {crawler_path} '{sources_json}'"
    print(f"Executing: {cmd}")
    
    result = subprocess.run(
        cmd,
        shell=True,
        check=True,
        capture_output=True,
        text=True
    )
    
    print("\nOutput:")
    print(result.stdout)
    
    if result.stderr:
        print("\nErrors:")
        print(result.stderr)
    
    # Parse JSON output
    try:
        output = json.loads(result.stdout)
        print("\nParsed JSON output:")
        print(f"Stories found: {len(output.get('stories', []))}")
        print(f"Errors: {len(output.get('errors', []))}")
    except json.JSONDecodeError:
        print("\nCouldn't parse JSON output")
        
except subprocess.CalledProcessError as e:
    print(f"\nError executing crawler: {e}")
    print(f"Exit code: {e.returncode}")
    if e.stdout:
        print(f"Output: {e.stdout}")
    if e.stderr:
        print(f"Error output: {e.stderr}")
    
except Exception as e:
    print(f"\nUnexpected error: {e}")
